File: android/src/main/kotlin/dev/asdevs/signalr_flutter/SignalR.kt
package dev.asdevs.signalr_flutter

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import microsoft.aspnet.signalr.client.*
import microsoft.aspnet.signalr.client.hubs.HubConnection
import microsoft.aspnet.signalr.client.hubs.HubProxy
import microsoft.aspnet.signalr.client.transport.LongPollingTransport
import microsoft.aspnet.signalr.client.transport.ServerSentEventsTransport
import android.util.Log


enum class CallMethod(val value: String) {
    ConnectToServer("connectToServer"),
    Reconnect("reconnect"),
    Stop("stop"),
    IsConnected("isConnected"),
    ListenToHubMethod("listenToHubMethod"),
    InvokeServerMethod("invokeServerMethod")
}

object SignalR {
    private val connections: MutableMap<String, HubConnection> = mutableMapOf()
    private val hubs: MutableMap<String, HubProxy> = mutableMapOf()
    val channels = mutableMapOf<String, MethodChannel>()

    fun connectToServer(id: String, url: String, hubName: String, queryString: String, headers: Map<String, String>, transport: Int, hubMethods: List<String>, result: Result) {

        if(connections.containsKey(id)) {
            result.error("Error", "Connection with this ID already exists.", null)
            return
        }


        try {
            val connection: HubConnection = if (queryString.isEmpty()) {
                HubConnection(url)
            } else {
                HubConnection(url, queryString, true, Logger { _: String, _: LogLevel -> })
            }

            if (headers.isNotEmpty()) {
                val cred = Credentials() { request ->
                    request.headers = headers
                }
                connection?.credentials = cred
            }
            val hub = connection!!.createHubProxy(hubName)!!

            hubMethods.forEach { methodName ->
                hub?.on(methodName, { res ->
                    Handler(Looper.getMainLooper()).post {
                        Log.d("SignalR2", "SignalR2 NewMsg")

                        SignalRFlutterPlugin.channel.invokeMethod("NewMessage", listOf(methodName, res))
                    }
                }, Any::class.java)
            }

            connection?.connected {
                Handler(Looper.getMainLooper()).post {
                    if(connection != null) {
                        Log.d("SignalR2", "SignalR2 Connected")
                        SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", "Connected")
                    }
                }
            }

            connection?.reconnected {
                Handler(Looper.getMainLooper()).post {
                    if(connection != null) {
                        Log.d("SignalR2", "SignalR2 Reconnected")

                        SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", "Reconnected")
                    }
                }
            }

            connection?.reconnecting {
                Handler(Looper.getMainLooper()).post {
                    if(connection != null) {
                        Log.d("SignalR2", "SignalR2 Reconnecting")

                        SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", "Reconnecting")
                    }
                }
            }

            connection?.closed {
                Handler(Looper.getMainLooper()).post {
                    if(connection != null) {
                        Log.d("SignalR2", "SignalR2 Closed")

                        SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", "Disconnected")
                    }
                }
            }

            connection?.connectionSlow {
                Handler(Looper.getMainLooper()).post {
                    Log.d("SignalR2", "SignalR2 Slow")

                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", "Slow")
                }
            }

            connection?.error { handler ->
                Handler(Looper.getMainLooper()).post {
                    Log.d("SignalR2", "SignalR2 Error " + handler.localizedMessage)

                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", "Error: " + handler.localizedMessage)
                }
            }

            when (transport) {
                1 -> connection?.start(ServerSentEventsTransport(connection?.logger))
                2 -> connection?.start(LongPollingTransport(connection?.logger))
                else -> {
                    connection?.start()
                }
            }

            result.success(true)
            connections[id] = connection
            hubs[id] = hub
        } catch (ex: Exception) {
            Log.d("SignalR2", "SignalR2 Error " + ex.localizedMessage)

            result.error("SignalR2 Error", ex.localizedMessage, null)
        }
    }

    fun reconnect(connectionId: String,result: Result) {
        try {
            connections[connectionId]?.start()
        } catch (ex: Exception) {
            result.error(ex.localizedMessage, ex.stackTrace.toString(), null)
        }
    }

    fun stop(id: String, result: Result) {
        try {
            connections[id]?.stop()
            connections.remove(id)
            hubs.remove(id)
        } catch (ex: Exception) {
            result.error(ex.localizedMessage, ex.stackTrace.toString(), null)
        }
    }

    fun isConnected(id: String, result: Result) {
        try {
            val connection = connections[id]
            if (connection != null) {
                when (connection.state) {
                    ConnectionState.Connected -> result.success(true)
                    else -> result.success(false)
                }
            } else {
                result.success(false)
            }
        } catch (ex: Exception) {
            result.error("Error", ex.localizedMessage, null)
        }
    }

    fun listenToHubMethod(connectionId: String, methodName: String, result: Result) {
        try {
            hubs[connectionId]?.on(methodName, { res ->
                println("SignalR3 - " + methodName)
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("NewMessage", listOf(methodName, res))
                }
            }, Any::class.java)
        } catch (ex: Exception) {
            println("SignalR3 ERROR")
            result.error("Error", ex.localizedMessage, null)
        }
    }

    fun invokeServerMethod(connectionId: String, methodName: String, args: List<Any>, result: Result) {
        try {
            val res: SignalRFuture<Any>? = hubs[connectionId]?.invoke(Any::class.java, methodName, *args.toTypedArray())

            res?.done { msg: Any? ->
                Handler(Looper.getMainLooper()).post {
                    result.success(msg)
                }
            }

            res?.onError { throwable ->
                Handler(Looper.getMainLooper()).post {
                    result.error("Error", throwable.localizedMessage, null)
                }
            }
        } catch (ex: Exception) {
            result.error("Error", ex.localizedMessage, null)
        }
    }
}

File: ios/Classes/SignalR.swift
//
//  SignalR.swift
//  signalR
//
//  Created by Ayon Das on 23/07/20.
//

import Foundation

enum CallMethod : String {
  case connectToServer, reconnect, stop, isConnected, invokeServerMethod, listenToHubMethod
}
class SignalRWrapper {

    private var hub: Hub!
    private var connection: SignalR!
    
    func connectToServer(arguments: [String: Any], result: @escaping FlutterResult) {
        guard let baseUrl = arguments["baseUrl"] as? String,
              let hubName = arguments["hubName"] as? String else {
            result(FlutterError(code: "Invalid Argument", message: "baseUrl or hubName missing", details: nil))
            return
        }
        
        connection = SignalR(baseUrl)
        
        let queryString = arguments["queryString"] as? String ?? ""
        if !queryString.isEmpty {
            let qs = queryString.components(separatedBy: "=")
            connection.queryString = [qs[0]:qs[1]]
        }
        
        if let transport = arguments["transport"] as? Int {
            switch transport {
            case 1:
                connection.transport = Transport.serverSentEvents
            case 2:
                connection.transport = Transport.longPolling
            default:
                break
            }
        }
        
        if let headers = arguments["headers"] as? [String: String], headers.count > 0 {
            connection.headers = headers
        }
        
        hub = connection.createHubProxy(hubName)
        
        if let hubMethods = arguments["hubMethods"] as? [String] {
            hubMethods.forEach { (methodName) in
                hub.on(methodName) { (args) in
                    SwiftSignalRFlutterPlugin.channel.invokeMethod("NewMessage", arguments: [methodName, args?[0]])
                }
            }
        }
        
        configureConnectionCallbacks()
        
        connection.start()
        result(true)
    }

    func reconnect(result: @escaping FlutterResult) {
        if let connection = self.connection {
            connection.connect()
        } else {
            result(FlutterError(code: "Error", message: "SignalR Connection not found or null", details: "Start SignalR connection first"))
        }
    }

    func stop(result: @escaping FlutterResult) {
        if let connection = self.connection {
            connection.stop()
        } else {
            result(FlutterError(code: "Error", message: "SignalR Connection not found or null", details: "Start SignalR connection first"))
        }
    }

    func isConnected(result: @escaping FlutterResult) {
        if let connection = self.connection {
            switch connection.state {
            case .connected:
                result(true)
            default:
                result(false)
            }
        } else {
            result(false)
        }
    }

    func listenToHubMethod(methodName: String, result: @escaping FlutterResult) {
        if let hub = self.hub {
            hub.on(methodName) { (args) in
//                SwiftSignalRFlutterPlugin.channel.invokeMethod("NewMessage", arguments: [methodName, args?[0]])
                guard let args = args, !args.isEmpty else {
                    SwiftSignalRFlutterPlugin.channel.invokeMethod("NewMessage", arguments: [methodName, "null or empty"])
                    return
                }
                
                let jsonData = try? JSONSerialization.data(withJSONObject: args[0], options: [])
                let jsonString = String(data: jsonData!, encoding: .utf8)
                SwiftSignalRFlutterPlugin.channel.invokeMethod("NewMessage", arguments: [methodName, jsonString ?? ""])
            }
        } else {
            result(FlutterError(code: "Error", message: "SignalR Connection not found or null", details: "Connect SignalR before listening a Hub method"))
        }
    }
    
    func tryToGetJsonData(input: Any?) -> Data? {
        guard let input = input else {
            return nil
        }
  
        if !JSONSerialization.isValidJSONObject(input) {
            return nil
        }

        do {
            let jsonData = try JSONSerialization.data(withJSONObject: input, options: [])
            return jsonData
        } catch {
            return nil
        }
    }

    func invokeServerMethod(arguments: [String: Any], result: @escaping FlutterResult) {
        guard let methodName = arguments["methodName"] as? String else {
            result(FlutterError(code: "Invalid Argument", message: "methodName missing", details: nil))
            return
        }
        
        do {
            if let hub = self.hub {
                try hub.invoke(methodName, arguments: arguments["arguments"] as? [Any], callback: { (res, error) in
                    if let error = error {
                        result(FlutterError(code: "Error", message: String(describing: error), details: nil))
                    } else {
                        let jsonData = self.tryToGetJsonData(input: res)

                        if jsonData == nil {
                            if res != nil {
                                let intResult = res as! Int
                                result(String(intResult))
                            } else {
                                result("")
                            }
                        } else {
                            let jsonString = String(data: jsonData!, encoding: .utf8)
                            result(jsonString ?? "")
                        }
                    }
                })
            } else {
                throw NSError.init(domain: "NullPointerException", code: 0, userInfo: [NSLocalizedDescriptionKey : "Hub is null. Initiate a connection first."])
            }
        } catch {
            result(FlutterError.init(code: "Error", message: error.localizedDescription, details: nil))
        }
    }
    
    private func configureConnectionCallbacks() {
        connection.starting = { [weak self] in
            print("SignalR Connecting. Current Status: \(String(describing: self?.connection.state.stringValue))")
            SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Connecting")
        }

        connection.reconnecting = { [weak self] in
            print("SignalR Reconnecting. Current Status: \(String(describing: self?.connection.state.stringValue))")
            SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Reconnecting")
        }

        connection.connected = { [weak self] in
            print("SignalR Connected. Connection ID: \(String(describing: self?.connection.connectionID))")
            SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Connected")
        }

        connection.reconnected = { [weak self] in
            print("SignalR Reconnected...")
            print("Connection ID: \(String(describing: self?.connection.connectionID))")
            SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Reconnected")
        }

        connection.disconnected = { [weak self] in
            print("SignalR Disconnected...")
            SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Disconnected")
        }

        connection.connectionSlow = {
            print("Connection slow...")
            SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Slow")
        }

        connection.error = { [weak self] error in
            print("Error: \(String(describing: error))")
            SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Error: \(String(describing: error))")
            if let source = error?["source"] as? String, source == "TimeoutException" {
                print("Connection timed out. Restarting...")
                self?.connection.start()
            }
        }
    }
}
